### PR TITLE
OSMD VexflowPatch: StaveRepetition, TO_CODA support and position fix.

### LINKED ISSUE #1255: OSMD VexflowPatch: StaveRepetition
add TO_CODA enum to type() and draw()
fix x-positioning for TO_CODA and DS_AL_CODA in drawSymbolText()
fix y-shift

### PR DESCRIPTION
**Pull request by rvilarl**
Fixes #1255 

There only visual differences on the two tests below (I have just placed the Gonville results). IT IS OBVIOUSLY WRONG.

Order is:
- diff
- reference
- current

**Stave Multiple_Staves_Volta**
Stave Multiple_Staves_Volta_Test Gonville
Stave Multiple_Staves_Volta_Test Gonville_Reference
Stave Multiple_Staves_Volta_Test Gonville_Current

**Stave Volta___Modifier_Measure**
Stave Volta___Modifier_Measure_Test Gonville
Stave Volta___Modifier_Measure_Test Gonville_Reference
Stave Volta___Modifier_Measure_Test Gonville_Current

### CODE REVIEW

**Comment by 0xfe** [comment]
This looks good to me, Rodrigo. Can you clean up the comments please?

**Comment by rvilarl** [comment]
> This looks good to me, Rodrigo. Can you clean up the comments please?

Do you mean remove them?

**Comment by 0xfe** [comment]
> > This looks good to me, Rodrigo. Can you clean up the comments please?
> 
> Do you mean remove them?

The parts where it says `VexflowPatch` -- no need for those.

**Comment by rvilarl** [comment]
> > > This looks good to me, Rodrigo. Can you clean up the comments please?
> > 
> > 
> > Do you mean remove them?
> 
> The parts where it says `VexflowPatch` -- no need for those.

Done, ready to merge.

**Comment by rvilarl** [comment]
@0xfe @sschmidTU this one seems to be a similar problem, does not it? What the element related to the text "a tempo"?

image

**Comment by rvilarl** [comment]
@0xfe @sschmidTU it seems that the fix of the x and y position is no longer required. It is already corrected in master. I will investigate this together with the `StaveTempo` above and I will adjust this PR.

**Comment by sschmidTU** [comment]
@rvilarl Current is the state of this PR, right? So there's a y positioning error with this PR, as seen in the first test (images).
Also, what did you mean with "This patch is rather introducing"?

**Comment by rvilarl** [comment]
> @rvilarl Current is the state of this PR, right? 

yes

> So there's a y positioning error with this PR, as seen in the first test (images). Also, what did you mean with "This patch is rather introducing"?

That the changes are introducing an error.
 
> And with a jsfiddle, do you mean reproducing the sample I provided in #1255 with a pure Vexflow implementation?

yes to include it in the set of tests

**Comment by rvilarl** [comment]
> > And with a jsfiddle, do you mean reproducing the sample I provided in #1255 with a pure Vexflow implementation?
> 
> yes to include it in the set of tests

@sschmidTU can you provide the pure Vexflow Implementation of the test?

**Comment by rvilarl** [comment]
I have refactor the StaveRepetion and its associated tests (the default 25 for x was set in the test) And I have added a test similar to the one mentioned by @sschmidTU in #1255. I belive that the results are good.

**Stave Volta___Modifier_Measure_Test Bravura**
Stave Volta___Modifier_Measure_Test Bravura

**Stave Multiple_Staves_Volta_Test Bravura**
Stave Multiple_Staves_Volta_Test Bravura

**NEW TEST: Stave Stave_Repetition__CODA__Positioning Bravura**
Stave Stave_Repetition__CODA__Positioning Bravura

**Comment by rvilarl** [comment]
@0xfe ready to review

File: src/staverepetition.ts [hunk 1]
```diff
@@ -121,24 +125,32 @@ export class Repetition extends StaveModifier {
...
-      text_x = symbol_x - +ctx.measureText(text).width - 12;
+    let text_x = 0;
+    let symbol_x = 0;
+    const modifierWidth = stave.getNoteStartX() - this.x;
+    switch (this.symbol_type) {
+      // To the left with symbol
+      case Repetition.type.CODA_LEFT:
+        // Offset Coda text to right of stave beginning
+        text_x = this.x + stave.getVerticalBarWidth();
+        symbol_x = text_x + ctx.measureText(text).width + 12;
```
Comment (0xfe) [inline src/staverepetition.ts:136]: Can we move this hardcoded `12` to the metrics file please?

File: src/staverepetition.ts [hunk 2]
```diff
@@ -121,24 +125,32 @@ export class Repetition extends StaveModifier {
...
+        break;
+      // To the right with symbol
+      default:
+        text_x = this.x + x + this.x_shift + stave.getWidth() - 5 - modifierWidth - ctx.measureText(text).width - 12;
+        symbol_x = text_x + ctx.measureText(text).width + 12;
+        break;
     }
 
-    const y = stave.getYForTopText(stave.getNumLines()) + this.y_shift;
+    const y = stave.getYForTopText(stave.getNumLines()) + this.y_shift + 25;
```
Comment (0xfe) [inline src/staverepetition.ts:153]: This hardcoded `25` too.

File: src/staverepetition.ts [hunk 3]
```diff
@@ -121,24 +125,32 @@ export class Repetition extends StaveModifier {
...
+      case Repetition.type.DC:
+      case Repetition.type.DC_AL_FINE:
+      case Repetition.type.DS:
+      case Repetition.type.DS_AL_FINE:
+      case Repetition.type.FINE:
+        text_x = this.x + x + this.x_shift + stave.getWidth() - 5 - modifierWidth - ctx.measureText(text).width;
+        break;
+      // To the right with symbol
+      default:
+        text_x = this.x + x + this.x_shift + stave.getWidth() - 5 - modifierWidth - ctx.measureText(text).width - 12;
```
Comment (0xfe) [inline src/staverepetition.ts:148]: This `-17` and the above `-5` too please.

File: tests/stave_tests.ts [hunk 4]
```diff
@@ -478,7 +543,7 @@ function drawVolta(options: TestOptions, contextBuilder: ContextBuilder): void {
   // bar 8 - juxtapose eighth measure
   const mm8 = new Stave(mm7.getWidth() + mm7.getX(), mm1.getY(), 60);
   mm8.setEndBarType(BarlineType.DOUBLE);
-  mm8.setRepetitionTypeRight(Repetition.type.DS_AL_CODA, 25);
+  mm8.setRepetitionType(Repetition.type.DS_AL_CODA, 0);
```
Comment (0xfe) [inline tests/stave_tests.ts:546]: Since the second parameter is always `0`, can we make it a default. Also, I think we should have a couple of tests for non-zero parameters.

**Review by 0xfe — COMMENTED** [review]
Thanks Rodrigo. Looks good in general. Mostly minor comments.

File: src/staverepetition.ts (hunk 1 above)
Comment (rvilarl) [reply src/staverepetition.ts:136]: done

File: src/staverepetition.ts (hunk 2 above)
Comment (rvilarl) [reply src/staverepetition.ts:153]: done

File: src/staverepetition.ts (hunk 3 above)
Comment (rvilarl) [reply src/staverepetition.ts:148]: Done

File: tests/stave_tests.ts (hunk 4 above)
Comment (rvilarl) [reply tests/stave_tests.ts:547]: Done

**Comment by rvilarl** [comment]
@0xfe done, two additional tests are there with yShift -10 and yShift +20 (below the Bravura version)

**Stave Stave_Repetition__CODA__Positioning___10_ Bravura**
Stave Stave_Repetition__CODA__Positioning___10_ Bravura
**Stave Stave_Repetition__CODA__Positioning___20_ Bravura**
Stave Stave_Repetition__CODA__Positioning___20_ Bravura

**Comment by rvilarl** [comment]
@0xfe ready to merge.

**Comment by 0xfe** [comment]
LGTM!
